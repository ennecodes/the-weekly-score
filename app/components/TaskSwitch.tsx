import { Dialog, DialogPanel, DialogTitle, Switch } from "@headlessui/react";
import { useState } from "react";

import { Task } from "./RoutineCard";


export const TaskSwitch = ({ task }: { task: Task }) => {
    const [isChecked, setIsChecked] = useState(false);
    const [isModalOpen, setIsModalOpen] = useState(false);

    const {id, name} = task;
    return (
      <>
        <span>{name}</span>
        <Switch
          onChange={(checked) => {
            if (checked) {
              setIsChecked(checked);
            } else {
              setIsModalOpen(true);
            }
          }}
          checked={isChecked}
          className="group inline-flex h-6 w-11 items-center rounded-full bg-gray-200 transition data-[checked]:bg-pink-600"
          data-testid={`task-${id}-switch`}
        >
          <span className="size-4 translate-x-1 rounded-full bg-white transition group-data-[checked]:translate-x-6" />
        </Switch>
        <Dialog
          open={isModalOpen}
          onClose={() => setIsModalOpen(false)}
          className="relative z-50"
        >
          <div className="fixed inset-0 flex w-screen items-center justify-center p-4">
            <DialogPanel className="max-w-sm space-y-4 border bg-white p-12">
              <DialogTitle className="font-bold">
                Undo task completion?
              </DialogTitle>
              <p>Are you sure you want to mark {name} as incomplete?</p>
              <div className="flex gap-4">
                <button onClick={() => setIsModalOpen(false)}>Cancel</button>
                <button
                  onClick={() => {
                    setIsChecked(false);
                    setIsModalOpen(false);
                  }}
                >
                  Mark Incomplete
                </button>
              </div>
            </DialogPanel>
          </div>
        </Dialog>
      </>
    );
  };
  