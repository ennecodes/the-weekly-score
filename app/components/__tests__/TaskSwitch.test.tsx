import { render, screen, fireEvent} from "@testing-library/react";

import { TaskSwitch } from "../TaskSwitch";

const testTask = {id: "1", name: "Test Task"};

describe("TaskSwitch", () => {
    it("renders correctly", () => {
        const { container } = render(<TaskSwitch task={testTask} />);
        expect(container).toMatchSnapshot();
    });
    it("displays the task name", () => {
        render(<TaskSwitch task={testTask} />);
                expect(screen.getByText("Test Task")).toBeInTheDocument();
    });
    it("allows the user to mark the task as complete", () => {
        render(<TaskSwitch task={testTask} />);        const switchElement = screen.getByRole("switch");
        fireEvent.click(switchElement);
        expect(switchElement).toBeChecked();
    });
    it("displays a modal to confirm marking the task as incomplete", () => {
        render(<TaskSwitch task={testTask} />);        const switchElement = screen.getByRole("switch");
        fireEvent.click(switchElement);
        fireEvent.click(switchElement);
        const cancelButton = screen.getByText("Cancel");
        const markIncompleteButton = screen.getByText("Mark Incomplete");
        expect(cancelButton).toBeInTheDocument();
        expect(markIncompleteButton).toBeInTheDocument();
    });
    it("closes the modal when the user clicks the cancel button", () => {
        render(<TaskSwitch task={testTask} />);        const switchElement = screen.getByRole("switch");
        fireEvent.click(switchElement);
        fireEvent.click(switchElement);
        const cancelButton = screen.getByText("Cancel");
        fireEvent.click(cancelButton);
        expect(screen.queryByText("Cancel")).not.toBeInTheDocument();
    });
    it("marks the task as incomplete when the user clicks the mark incomplete button", () => {
        render(<TaskSwitch task={testTask} />);
        const switchElement = screen.getByRole("switch");
        fireEvent.click(switchElement);
        fireEvent.click(switchElement);
        const markIncompleteButton = screen.getByText("Mark Incomplete");
        fireEvent.click(markIncompleteButton);
        expect(switchElement).not.toBeChecked();
    });

});