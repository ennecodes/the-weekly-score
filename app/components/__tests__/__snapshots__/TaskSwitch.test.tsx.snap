// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TaskSwitch > renders correctly 1`] = `
<div>
  <span>
    Test Task
  </span>
  <button
    aria-checked="false"
    class="group inline-flex h-6 w-11 items-center rounded-full bg-gray-200 transition data-[checked]:bg-pink-600"
    data-headlessui-state=""
    data-testid="task-1-switch"
    id="headlessui-switch-:r0:"
    role="switch"
    tabindex="0"
    type="button"
  >
    <span
      class="size-4 translate-x-1 rounded-full bg-white transition group-data-[checked]:translate-x-6"
    />
  </button>
</div>
`;
